types/src/network_address.rs

```diff
@@ -0,0 +1,645 @@
+            }
+        }
+
+        #[derive(Deserialize)]
+        #[serde(rename = "NetworkAddress")]
+        struct DeserializeWrapper(Vec<Protocol>);
+
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_str(StringVisitor)
+        } else {
+            let wrapper = DeserializeWrapper::deserialize(deserializer)?;
+            let addr = NetworkAddress::try_from(wrapper.0).map_err(de::Error::custom)?;
+            Ok(addr)
+        }
+    }
+}
+
+#[cfg(any(test, feature = "fuzzing"))]
+impl Arbitrary for NetworkAddress {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        vec(any::<Protocol>(), 1..10)
+            .prop_map(NetworkAddress::new)
+            .boxed()
+    }
+}
+
+//////////////
+// Protocol //
+//////////////
+
+impl fmt::Display for Protocol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use self::Protocol::*;
+        match self {
+            Ip4(addr) => write!(f, "/ip4/{}", addr),
+            Ip6(addr) => write!(f, "/ip6/{}", addr),
+            Dns(domain) => write!(f, "/dns/{}", domain),
+            Dns4(domain) => write!(f, "/dns4/{}", domain),
+            Dns6(domain) => write!(f, "/dns6/{}", domain),
+            Tcp(port) => write!(f, "/tcp/{}", port),
+            Memory(port) => write!(f, "/memory/{}", port),
+            NoiseIk(pubkey) => write!(
+                f,
+                "/ln-noise-ik/{}",
+                pubkey
+                    .to_encoded_string()
+                    .expect("ValidKeyStringExt::to_encoded_string is infallible")
+            ),
+            Handshake(version) => write!(f, "/ln-handshake/{}", version),
+        }
+    }
+}
+
+/////////////
+// DnsName //
+/////////////
+
+impl DnsName {
+    pub fn is_valid(s: &str) -> Result<(), ParseError> {
+        if s.is_empty() {
+            Err(ParseError::EmptyDnsNameString)
+        } else if s.as_bytes().len() > MAX_DNS_NAME_SIZE {
+            Err(ParseError::DnsNameTooLong(s.as_bytes().len()))
+        } else if s.contains('/') {
+            Err(ParseError::InvalidDnsNameCharacter)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Into<String> for DnsName {
+    fn into(self) -> String {
+        self.0
+    }
+}
+
+impl AsRef<str> for DnsName {
+    fn as_ref(&self) -> &str {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<String> for DnsName {
+    type Error = ParseError;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        DnsName::is_valid(s.as_str()).map(|_| DnsName(s))
+    }
+}
+
+impl FromStr for DnsName {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        DnsName::is_valid(s).map(|_| DnsName(s.to_owned()))
+    }
+}
+
+impl fmt::Display for DnsName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Serialize for DnsName {
```

Comment:
Do these need to be hand implemented? seems like you should be able to just use the derives for these simple wrappers

Reply:
oh cool, looks like I can actually use the `#[serde(try_from = "String", into = "String")]`. good catch!